protocol/superchainerc20/convert.md

```diff
@@ -0,0 +1,196 @@
+  require(isValidPair(_legacyAddr, _superAddr), "Invalid address pair");
+
+  IERC20(_superAddr).burn(msg.sender, _amount);
+  IERC20(_legacyAddr).mint(msg.sender, _amount);
+}
+```
+
+The corresponding `SuperchainERC20` will need to allow the `L2StandardBridge` to mint and burn tokens.
+
+
+>💡 Note:
+We considered placing the `convert()` function directly in the `SuperchainERC20`, similar to how WETH manages deposits and withdrawals. However, we believe including them in the `L2StandardBridge` makes more sense. This is because:
+>- It makes the token leaner. Users will interact with `convert()` only a few times and in a context similar to deposit and withdraw.
+>- The `SuperchainERC20` representation for migrated liquidity will be as close as possible to newly deployed `SuperchainERC20`s.
+>- The `L2StandardBridge` needs to be involved because of burn and mint. Having the functionality in the token would have added additional external calls.
+
+
+### Access Control
+
+To check input addresses are valid, we will need to:
+
+- Check `SuperchainERC20` is valid: This depends on how the Factory is built.
+    - Store addresses: The simplest method is to store in the Factory a mapping that tracks the address of the corresponding `SuperchainERC20` for each L1 token (and other factors, such as decimals).
+    - Deterministic address: The Factory address is known, and so is the L1 token address, so the `SuperchainERC20` address can potentially be predicted. This heavily depends on the final choice of salt.
+- **Check `OptimismMintableERC20` is valid**: This involves verifying that the address was deployed from a valid factory. It is still undecided which Factories should be whitelisted. Possible approaches include:
+    - **Use a Registry**: An entity can update the registry at inception, and whitelisted factories can update it upon deployment. The Convert function will check mappings in this registry to allow a pair.
+    - **Use deterministic address**: This is only possible for some factories.
+        - The [new factory](https://github.com/ethereum-optimism/optimism/blob/v1.1.4/packages/contracts-bedrock/src/universal/OptimismMintableERC20Factory.sol) uses CREATE2 and will have a single representation. Deterministic address computation can verify the validity of the `OptimismMintableERC20`.
+        - The [old factory](https://github.com/ethereum-optimism/optimism-legacy/blob/develop/packages/contracts-bedrock/contracts/universal/OptimismMintableERC20Factory.sol) uses CREATE, meaning that the only possible approach is to iterate through the potential nonces. The maximum used nonce is not available in the contract, so the best approach would be to have some kind of registry (single-time update).
+- **Check `SuperchainERC20` address corresponds to the correct L1 address**: This is verified during the same validity check of the `SuperchainERC20`.
+
+A possible implementation would look like this:
+
+```solidity
+
```

Comment:
I am leaning towards the idea of adding a storage key/value pair to the factory for deployed addresses. We could:
- Update the factory to set storage after doing a deployment
- Do manual state surgery to fill in the storage slots on OPM + Base using an atomic upgrade to `StorageSetter` that sets values based on the offchain token list.

This would be a one time action that moves the token list on chain, we would not need to manually update the tokenlist over time. This would work well for chains like Base where they didn't use a canonical factory for some time. Then new chains don't need to worry about this storage setting and we can not need to add extra contracts to the mix.